include/deal.II/base/big_mpi_compat.h

```diff
@@ -0,0 +1,331 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2022 - 2022 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+/**
+ *
+ * BigMPICompat - a tiny MPI 4.x compatibility library
+ *
+ * Release under MIT at https://github.com/tjhei/BigMPICompat
+ */
+#ifndef BIG_MPI_COMPAT_H
+#define BIG_MPI_COMPAT_H
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_MPI
+
+#  include <mpi.h>
+#  ifndef MPI_VERSION
+#    error "Your MPI implementation does not define MPI_VERSION!"
+#  endif
+
+#  include <iostream>
+
+
+#  if MPI_VERSION < 3
+#    error "BigMPICompat requires at least MPI 3.0"
+#  endif
+
+/**
+ * This namespace contains symbols related to the BigMPICompat library
+ * to support large MPI routines on MPI implementations that implement
+ * version 3.x of the standard.
+ */
+namespace BigMPICompat
+{
+  static constexpr MPI_Count mpi_max_count = (1ULL << 31);
+
+} // namespace BigMPICompat
+
+#  if MPI_VERSION >= 4
+
+#  else
+
+/**
+ * Create a contiguous type of (possibly large) @p count.
+ *
+ * See the MPI 4.x standard for details.
+ */
+inline int
+MPI_Type_contiguous_c(MPI_Count     count,
+                      MPI_Datatype  oldtype,
+                      MPI_Datatype *newtype)
+{
+  if (count <= BigMPICompat::mpi_max_count)
+    return MPI_Type_contiguous(count, oldtype, newtype);
+  else
+    {
+      int             ierr;
+      const MPI_Count max_signed_int = (1U << 31) - 1;
+
+      MPI_Count size_old;
+      ierr = MPI_Type_size_x(oldtype, &size_old);
+
+      MPI_Count n_chunks          = count / max_signed_int;
+      MPI_Count n_bytes_remainder = count % max_signed_int;
+
+      MPI_Datatype chunks;
+      ierr = MPI_Type_vector(
+        n_chunks, max_signed_int, max_signed_int, oldtype, &chunks);
+      if (ierr != MPI_SUCCESS)
+        {
+          std::cerr << "MPI_Type_vector() call failed!" << std::endl;
```

Comment:
Could you take out the output here? If you want to be compatible with MPI, you should only return the error code. (Though there is probably little harm to leaving it in, it's only going to trigger in case of error anyway. The only harm I can think of is if the call triggers on 10,000 processors and you get a GB of error messages :-) )